src/object.rs

```diff
@@ -714,4 +717,56 @@ impl<T: IsA<Object> + SetValue> ObjectExt for T {
             }
         }
     }
+
+    fn downgrade(&self) -> WeakRef<T> {
+        unsafe {
+            let w = WeakRef(Box::new(mem::zeroed()), PhantomData);
+            gobject_ffi::g_weak_ref_init(mut_override(&*w.0), self.to_glib_none().0);
+            w
+        }
+
+    }
+}
+
+pub struct WeakRef<T: IsA<Object> + ?Sized>(Box<gobject_ffi::GWeakRef>, PhantomData<*const T>);
```

Comment:
Has to be a Box because GWeakRef remembers the address of this, and otherwise it would remember an "unstable" address (it is changing when moving the WeakRef around) and use it later (when the real object is destroyed).